### PR TITLE
ES|QL: Improve support for TEXT fields in functions

### PR DESCRIPTION
**Pull request by luigidellaquila**
Related to https://github.com/elastic/elasticsearch/issues/105379

ESQL functions that support KEYWORD fields, typically also support TEXT fields.
A few functions, indentified by https://github.com/elastic/elasticsearch/issues/105379, did not respect this convention.

This PR adds support for TEXT fields (where KEYWORD fields were allowed) for:
- split()
- cidr_match()
- date_format()
- date_parse()
- date_extract()

It also adds proper unit tests for `cidr_match()` and `date_format()`

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Split.java
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Split.java:1]: Shouldn't we have the same type check for both arguments here, given `right()` can be a field too?
Looking at it, we don't have any tests with the right argument being a field (or expression with one) -- it'd be good if we had one (looking at Marco's example).

**Review by bpintea — APPROVED** [review]
With some nits, it lgtm.

**Review by astefan — APPROVED** [review]
Except what @bpintea said, it LGTM.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateFormatTests.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import static org.hamcrest.Matchers.equalTo;
+
+public class DateFormatTests extends AbstractScalarFunctionTestCase {
+    public DateFormatTests(@Name("TestCase") Supplier<TestCaseSupplier.TestCase> testCaseSupplier) {
+        this.testCase = testCaseSupplier.get();
+    }
+
+    @ParametersFactory
+    public static Iterable<Object[]> parameters() {
+        return parameterSuppliersFromTypedData(
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/date/DateFormatTests.java:35]: It seems the function test infrastructure does not tolerate very well the fact that DateFormat accepts different types based on the fact that one of the two params is null (ie. `date_format(pattern, date)` vs `date_format(date)`).
If you don't mind, I'll try to address the problem in a follow-up PR, as it doesn't seem trivial.

I took a chance to add `errorsForCasesWithoutExamples()` to CIDRMatchTests though.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Split.java
Comment (luigidellaquila) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Split.java:1]: In ESQL it doesn't really make a big difference: all strings are considered exact, also TEXT, so isString() and isStringAndExact() are practically equivalent.
Anyway, I changed it to make the code more clear and consistent

**Pull request merged**